### ISSUE TITLE
WebSockets for notifications

### ISSUE THREAD

**Issue by elijah-wright** [bluesky-social/atproto#1875]
one of the things I noticed using Bluesky is that it polls for notifications, at least on the bsky.app website, with no cache policy. obviously, that's not instantaneous and constant polling puts more work on the server than what's actually necessary

from what I've seen, long polling is a good option, but it's almost always a fallback if WebSockets don't work (and I'm not suggesting server-side events because there's no guarantee that a packet is received on the client's end if not configured properly, like a proxy). I read something about event streams in the ATProto docs but I wasn't sure how that would work with this

one idea for consideration is using Redis to subscribe to a notifications channel with authentication using the Postgres database. I'm not sure how many connections or how much throughput could be supported, so more work might be needed there. in theory, if there's too much traffic for one streaming server, more could be added with PgBouncer if necessary

**Comment by bnewbold** [comment]
We use WebSockets already in the protocol. Streams can be specified using the lexicon schema language, details at https://atproto.com/specs/event-stream

The most notable current use is the firehose coming from the relay. You can see an example use of this feed at https://firesky.tv/

So an application or extension could already specify a stream for client notifications (eg, from appview to client). Though now that I think of it, how that would work with proxying via the PDS isn't defined, but shouldn't be too tricky to figure out.

As to if or when we roll that out for the app.bsky application, i'm not sure.

**Comment by elijah-wright** [comment]
yeah, I'm only talking about implementing this on bsky.app. I can develop it if necessary

**Comment by elijah-wright** [comment]
I read up on the Event Stream specification but I'm still not understanding it well
